### ISSUE TITLE
stix map reprojection

### ISSUE THREAD

**Issue by samaloney** [samaloney/stixpy#36]
> Thanks to the fundamental work of Erica (in cc), we built a python class that is a kind of a wrapper around reproject_to() in sunpy. The advantages of this wrapper are the following:
> 
> easy to use: with two lines, the user can reproject the map (three lines for a series of maps). The default is as seen from Solar Orbiter. This is good for over-plotting easily STIX images on top of reprojected AIA maps.
> it works for both full-disk maps and with cutouts. For a series of STIX images, it is good to have a series of reprojected AIA maps. To reproject a series of cutouts, it takes only a few seconds.
> if the user wants to store the reprojected map as a FITS file, the newly generated FITS (for both full-disks and cutouts) is compatible with the standard fits2map IDL reader. This is how we currently reproject the maps that we then import in IDL when comparing STIX images with AIA.
> Do you think would it be useful to include it into stixpy?

**Comment by hayesla** [comment]
I think this sounds great, and something that is needed!

 Is it a quicker version that what is used in `reproject_to` so that it is faster?

**Comment by elastufka** [comment]
@hayesla yes, it's (usually) quicker because it will reproject the cutout to its correct spatial extent in the desired observer frame, instead of always projecting it onto a hemisphere, which in the case of cutouts will always be larger than necessary.  This could eventually be built into Sunpy itself, but the issue of proper pixel scaling between maps (for example, AIA to STEREO) would need to be resolved first. 

Also as I was discussing with @afbattaglia, the SPICE kernel dependency can be removed by getting the trajectory information from stixdcpy. (other options are the  using spiceypy or the old, working version of heliopy, which also download the kernel just behind-the-scenes, or heliopy's replacement astrospice which as I understand will do the same thing once it has more functionality to it)

**Comment by samaloney** [comment]
Sounds great, about the SPICE the issue is always going to be trying to make sure he same version used in processing the data and is also used doing the reprojection. But that's a separate issue so for this I'd assume you have access to the the spacecraft location and make that the API 

For [example](https://stixcore.readthedocs.io/en/latest/api/stixcore.ephemeris.manager.Spice.html#stixcore.ephemeris.manager.Spice.get_position)

**Comment by afbattaglia** [comment]
Great! Thanks @samaloney and @hayesla 
@elastufka I think that next time we meet at FHNW we can discuss how to proceed. I still have some tests that I would like to do but this week I'm pretty busy and next week I'll be on holiday. If you agree, we can discuss it in two weeks or so

**Comment by elastufka** [comment]
Hi @samaloney , 

Following discussion today with @afbattaglia and the rest of the team, this is something we would like to have available soon. 

There are a few ways of going about it, depending on how much work is required by the user. 

1) combination of SunPy and local SPICE kernel or (old, working version of) heliopy (can eventually be replaced with astrospice once that gains functionality). Users needs to know how to get trajectory information and construct header
 
2) existing (but optimized for submaps) version of map_reprojection.py. Handles the SPICE kernel/trajectory information internally. Limited to projecting other maps into SOLO reference frame.

3) existing code used for reprojecting any map or submap to any other reference frame. Handles the SPICE kernel/trajectory information internally via heliopy (or drms in the case of AIA). User only needs to specify correct code for desired reference frame. [available here but currently private](https://github.com/afbattaglia/STIX-science) 

I would suggest at the very least adding a detailed tutorial for reprojection and optimizing map_reprojection.py to handle submaps. Ideally such an optimization will be accepted for inclusion in SunPy as well. I can work on this right away.

I think it would be good to also AIA trajectory information via drms (the difference between the Earth-observer and AIA-observer can make quite a difference near the solar limb).

Personally I am in favor of the full implementation, ie option 3 with doing all the kernel/trajectory operations under the hood. However maybe this warrants some more discussion as to what exactly stixpy should offer to users, since reprojecting STIX images for example can be easily misinterpreted.

**Comment by samaloney** [comment]
So I'm also in favour of 3 albeit in a a different guise. As part of STIXCore we will be producing auxiliary files which will have the position, attitude and aspect solution every 64s which should be sufficient? for some examples see http://pub099.cs.technik.fhnw.ch/data/aux/

There are issues around SPICE kernels e.g. size and synchronisation to start. In order to do the reprojections correctly needs the as-flown trajectory and as-flown pointings which are ~100s MB and growing. For example AFAIK  the deprecated heliopy and or new astro astrospice package only download/use the planned orbit and attitudes. For most accurate results the aspect solution is also needed.

Overall this should have the same API and functionally as sunpy map reprojection so I'd suggest getting a PR open and let get this moving forward

**Comment by elastufka** [comment]
I'm not worried about the trajectory info for STIX. In option 3 I am mostly talking about trajectory information for other spacecraft or celestial bodies.

The SunPy method right now is closer to option 2. The user needs to provide a header or WCS, most easily done by having a map from the instrument in which the original map will be reprojected to. Option 3 means something like:

```python
original_map=sunpy.map.Map('aia_fitsfile.fits')
reprojected_map_solo=stixpy.map_reproject(original_map, output_frame='SOLO')
reprojected_map_stereo=stixpy.map_reproject(original_map, output_frame='STEREO AHEAD')
reprojected_map_mars=stixpy.map_reproject(original_map, output_frame='MARS')
reprojected_map_anyframe=stixpy.map_reproject(original_map, output_frame='ANY_SPICE_CODE')
```

This is similar what Andrea and I are using now.

Re: old heliopy - (some) kernels there can be specified to be as-flown or predicted, and they will have different names, ie _solo_ vs _solo_pred_ . In spiceypy I believe it is also possible to specify to use as-flown, but it's a bit less obvious how.

**Comment by samaloney** [comment]
Ah ok I see what you mean I think there's two distinct use cases 1) Map to Map and 2) Map to arbitrary observer. For 1 in the API or interface is already clear the two maps. stixpy shouldn't and won't care if the header info is correct in Maps (e.g. if a user doesn't use aiapy to prep aia data and update the position e.t.c. For the second we already have something but it could be expanded upon and improved.

The issue with the kernels is support but its downloading the right one with the relevant information I mean we could always use the latest but that a few 100mbs to download each time. So I think using our aux files would be the best way to go that way the our `STIXMap`s will have the correct was info.

**Comment by elastufka** [comment]
Okay. This still doesn't entirely answer my question, which is do we want to include Map to arbitrary observer in stixpy. Can you elaborate on, "For the second we already have something but it could be expanded upon and improved." From  what I can gather, you only have SOLO and Earth trajectory information. 

I'll submit a PR including getting the AIA location information via drms (probably number 1 use case of reprojection is STIX-AIA and vice versa) and then we can go from there.
